src/cargo/core/resolver/mod.rs

```diff
@@ -884,80 +884,205 @@ fn generalize_conflicting(
         // and let `backtrack_critical_id` figure this out.
         return None;
     }
+
+    // we will need to know the range of versions we can use
+    let our_candidates = registry
+        .query(dep)
+        .expect("an already used dep now error!?");
+
+    if our_candidates.is_empty() {
+        return None;
+    }
+
+    let our_activation_keys: HashSet<_> = our_candidates
+        .iter()
+        .map(|c| c.summary.package_id().as_activations_key())
+        .collect();
+
+    // if our dep only matches one semver range then we can fast path any other dep
+    // that also targets that semver range and has no overlap
+    let our_activation_key = if our_activation_keys.len() == 1 {
+        our_activation_keys.iter().next().cloned()
+    } else {
+        None
+    };
+
+    let our_links: Option<HashSet<_>> = our_candidates.iter().map(|c| c.summary.links()).collect();
```

Comment:
This might be more clear to use `filter_map` and then collect directly into `HashSet` instead of `Option<HashSet>` perhaps?